### PR TITLE
SONARJAVA-547 Remove usage of AstNode to get tokens and constructors using AstNode children

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/model/statement/ForStatementTreeImpl.java [hunk 1]
```diff
@@ -33,21 +31,34 @@
...
+  private final InternalSyntaxToken firstSemicolonToken;
   @Nullable
   private final ExpressionTree condition;
+  private final InternalSyntaxToken secondSemicolonToken;
   private final List<StatementTree> update;
+  private final InternalSyntaxToken closeParenToken;
   private final StatementTree statement;
 
-  public ForStatementTreeImpl(List<StatementTree> initializer, @Nullable ExpressionTree condition, List<StatementTree> update, StatementTree statement) {
+  public ForStatementTreeImpl(InternalSyntaxToken forKeyword, InternalSyntaxToken openParenToken, List<StatementTree> initializer,
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/statement/ForStatementTreeImpl.java]: MAJOR Method has 9 parameters, which is greater than 7 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)

File: java-squid/src/main/java/org/sonar/java/model/statement/ForEachStatementImpl.java [hunk 2]
```diff
@@ -37,14 +35,23 @@
...
+  private final InternalSyntaxToken forKeyword;
+  private final InternalSyntaxToken openParenToken;
   private final VariableTree variable;
+  private final InternalSyntaxToken colonToken;
   private final ExpressionTree expression;
+  private final InternalSyntaxToken closeParenToken;
   private final StatementTree statement;
 
-  public ForEachStatementImpl(VariableTree variable, ExpressionTree expression, StatementTree statement, AstNode... children) {
+  public ForEachStatementImpl(InternalSyntaxToken forKeyword, InternalSyntaxToken openParenToken, VariableTree variable, InternalSyntaxToken colonToken,
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/statement/ForEachStatementImpl.java]: MAJOR Method has 8 parameters, which is greater than 7 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

**Comment by benzonico** [comment]
As discussed, can you remove AstNode varargs where achievable.

File: java-squid/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java [hunk 3]
```diff
@@ -116,6 +125,21 @@ public TryStatementTreeImpl completeStandardTry(InternalSyntaxToken tryToken, Bl
     return this;
   }
 
+  public TryStatementTreeImpl completeTryWithResources(InternalSyntaxToken tryToken, InternalSyntaxToken openParenToken, ResourceListTreeImpl resources,
+    InternalSyntaxToken closeParenToken, BlockTreeImpl block, List<CatchTreeImpl> catches) {
+    this.tryToken = tryToken;
+    this.openParenToken = openParenToken;
+    this.resources = (List) resources;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java]: MINOR Remove this unnecessary cast to "List". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1905)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 major
* 1 minor

File: java-squid/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java [hunk 4]
```diff
@@ -101,10 +107,30 @@ public TryStatementTreeImpl(
     for (CatchTreeImpl catchTree : catches) {
       addChild(catchTree);
     }
+  }
 
-    if (finallyBlock != null) {
-      addChild(finallyBlock);
+  private List<VariableTree> getResources(ResourceListTreeImpl resources) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java]: MAJOR Make "getResources" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: java-squid/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java [hunk 5]
```diff
@@ -101,10 +107,30 @@ public TryStatementTreeImpl(
...
+    ImmutableList.Builder<VariableTree> resourcesBuilder = ImmutableList.builder();
+    for (VariableTreeImpl resource : resources) {
+      resourcesBuilder.add(resource);
     }
+    return resourcesBuilder.build();
+  }
+
+  private List<CatchTree> getCatches(List<CatchTreeImpl> catches) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java]: MAJOR Make "getCatches" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 6]
```diff
@@ -1651,6 +1660,7 @@ public TypeCastExpressionTreeImpl newClassCastExpression(TypeTree type, Optional
 
     List<AstNode> children = Lists.newArrayList();
     children.add((AstNode) type);
+    // TODO SONARJAVA-1139
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not part of this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/95c5a693507566b5bc922697dff16df88e7dab4d/java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java#L2124) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 7]
```diff
@@ -1651,6 +1659,7 @@ public TypeCastExpressionTreeImpl newClassCastExpression(TypeTree type, Optional
 
     List<AstNode> children = Lists.newArrayList();
     children.add((AstNode) type);
+    // TODO SONARJAVA-1139 bound should be present in castTree
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 1 major
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not part of this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/5666a8fcbd0bdcc69b36492d2250944798f9ef26/java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java#L2123) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 8]
```diff
@@ -2091,16 +2058,9 @@ public ExpressionTree newIdentifierOrMethodInvocation(Optional<TypeArgumentListT
...
-      children.add(nestedDimensions);
-    }
-    children.add(dotTokenAstNode);
-    children.add(identifier);
 
-    return new MemberSelectExpressionTreeImpl(nestedDimensions, dotToken, identifier, children);
+    return new MemberSelectExpressionTreeImpl(nestedDimensions, dotToken, identifier);
   }
 
   private ExpressionTree applySelectors(ExpressionTree primary, Optional<List<Tuple<Optional<InternalSyntaxToken>, ExpressionTree>>> selectors) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Make "applySelectors" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by benzonico** [comment]
LGTM

**Pull request merged**